rubullet/src/types.rs

```diff
@@ -0,0 +1,1128 @@
+            world_angular_velocity: m_world_angular_velocity,
+        }
+    }
+}
+
+fn position_orientation_to_isometry(position: [f64; 3], orientation: [f64; 4]) -> Isometry3<f64> {
+    Isometry3::<f64>::from_parts(
+        Translation3::from(Vector3::from_column_slice(&position)),
+        UnitQuaternion::from_quaternion(Quaternion::from_parts(
+            orientation[3],
+            Vector3::from_column_slice(&orientation[0..3]),
+        )),
+    )
+}
+fn combined_position_orientation_array_to_isometry(combined: [f64; 7]) -> Isometry3<f64> {
+    let position = [combined[0], combined[1], combined[2]];
+    let orientation = [combined[3], combined[4], combined[5], combined[6]];
+    position_orientation_to_isometry(position, orientation)
+}
+
+/// VisualShape options are for the [create_visual_shape](`crate::PhysicsClient::create_visual_shape`)
+/// function to specify additional options like the color.
+pub struct VisualShapeOptions {
+    /// offset of the shape with respect to the link frame
+    pub frame_offset: Isometry3<f64>,
+    /// color components for red, green, blue and alpha, each in range [0,1]
+    pub rgba_colors: [f64; 4],
+    /// specular reflection color, red, green, blue components in range [0,1]
+    pub specular_colors: [f64; 3],
+    /// Additional flags. Currently not used
+    pub flags: Option<i32>,
+}
+impl Default for VisualShapeOptions {
+    fn default() -> Self {
+        VisualShapeOptions {
+            frame_offset: Isometry3::translation(0., 0., 0.),
+            rgba_colors: [1.; 4],
+            specular_colors: [1.; 3],
+            flags: None,
+        }
+    }
+}
+/// Collision shape which can be put
+/// the [create_collision_shape](`crate::PhysicsClient::create_collision_shape`) method
+pub enum GeometricCollisionShape {
+    /// A Sphere determined by the radius in meter
+    Sphere {
+        /// radius in meter
+        radius: f64,
+    },
+    /// A Cuboid
+    Box {
+        /// [x,y,z] lengths starting from the middle of the box.
+        /// For example [0.5,0.5,0.5] is a unit cube.
+        half_extents: [f64; 3],
+    },
+    /// Like a cylinder but with a half sphere on each end. The total length of a capsule is
+    /// length + 2 * radius.
+    Capsule {
+        /// radius of the cylindric part of the capsule in meter.
+        radius: f64,
+        /// height of the cylindric part in meter. The half spheres are put on top on that
+        height: f64,
+    },
+    /// A Cylinder
+    Cylinder {
+        /// radius in meter
+        radius: f64,
+        /// height in meter
+        height: f64,
+    },
+    /// A Plane.
+    Plane {
+        /// normal of the plane.
+        plane_normal: [f64; 3],
+    },
+    /// Load a .obj (Wavefront) file. Will create convex hulls for each object.
+    MeshFile {
+        /// Path to the .obj file as String.
+        filename: String,
+        /// Scaling of the Mesh. Use [1.;3] for original scaling.
+        mesh_scale: [f64; 3],
```

Comment:
Arguably a `Vector3`

Reply:
Same as the half extents. There is also a problem with the Mesh enums. They have vertices and normals which should obviously be Point3 and Vector3. However, it would be inconvenient to create them this way. At some point in time there should be builder for those to make use of the generics. I would open an issue for that if you think it is a good idea

Reply:
My comments for the half extents apply here as well. In short: Using a vector is what [Bullet itself does][1] but I won't be upset if you don't agree.

[1]: https://pybullet.org/Bullet/BulletFull/classbtScaledBvhTriangleMeshShape.html

Reply:
You convinced me. However typing `Vector3::ones()` every time is a bit annoying. How about renaming the argument to `mesh_scaling` and make it an `Option<Vector3<f64>>` and None would be the original scaling?

Reply:
I very much support that.

Reply:
done